src/backends/wasm_js.rs

```diff
@@ -7,53 +7,68 @@ pub use crate::util::{inner_u32, inner_u64};
-    let buf = Uint8Array::new_with_length(CRYPTO_BUFFER_SIZE.into());
-    for chunk in dest.chunks_mut(CRYPTO_BUFFER_SIZE.into()) {
-        let chunk_len: u32 = chunk
-            .len()
-            .try_into()
-            .expect("chunk length is bounded by CRYPTO_BUFFER_SIZE");
-        // The chunk can be smaller than buf's length, so we call to
-        // JS to create a smaller view of buf without allocation.
-        let sub_buf = buf.subarray(0, chunk_len);
-
-        if crypto.get_random_values(&sub_buf).is_err() {
-            return Err(Error::WEB_GET_RANDOM_VALUES);
+    CRYPTO.with(|crypto| {
+        let crypto = crypto.as_ref().ok_or(Error::WEB_CRYPTO)?;
+
+        #[cfg(not(target_feature = "atomics"))]
+        {
+            for chunk in dest.chunks_mut(MAX_BUFFER_SIZE.into()) {
+                if crypto.get_random_values(chunk).is_err() {
+                    return Err(Error::WEB_GET_RANDOM_VALUES);
+                }
+            }
         }
+        #[cfg(target_feature = "atomics")]
+        {
+            // getRandomValues does not work with all types of WASM memory,
+            // so we initially write to browser memory to avoid exceptions.
+            let buf = Uint8Array::new_with_length(MAX_BUFFER_SIZE.into());
+            for chunk in dest.chunks_mut(MAX_BUFFER_SIZE.into()) {
+                let chunk_len: u32 = chunk
+                    .len()
+                    .try_into()
+                    .expect("chunk length is bounded by MAX_BUFFER_SIZE");
+                // The chunk can be smaller than buf's length, so we call to
+                // JS to create a smaller view of buf without allocation.
+                let sub_buf = if chunk_len == u32::from(MAX_BUFFER_SIZE) {
+                    buf.clone()
+                } else {
+                    buf.subarray(0, chunk_len)
+                };
+
+                if crypto.get_random_values(&sub_buf).is_err() {
+                    return Err(Error::WEB_GET_RANDOM_VALUES);
+                }
 
-        // SAFETY: `sub_buf`'s length is the same length as `chunk`
-        unsafe { sub_buf.raw_copy_to_ptr(chunk.as_mut_ptr().cast::<u8>()) };
-    }
-    Ok(())
+                // SAFETY: `sub_buf`'s length is the same length as `chunk`
+                unsafe { sub_buf.raw_copy_to_ptr(chunk.as_mut_ptr().cast::<u8>()) };
+            }
+        }
+        Ok(())
+    })
 }
 
 #[wasm_bindgen]
 extern "C" {
-    // Return type of js_sys::global()
-    type Global;
     // Web Crypto API: Crypto interface (https://www.w3.org/TR/WebCryptoAPI/)
     type Crypto;
-    // Getters for the Crypto API
-    #[wasm_bindgen(method, getter)]
-    fn crypto(this: &Global) -> Crypto;
+    // Holds the global `Crypto` object.
+    #[wasm_bindgen(thread_local_v2, js_name = crypto)]
+    static CRYPTO: Option<Crypto>;
     // Crypto.getRandomValues()
+    #[cfg(not(target_feature = "atomics"))]
+    #[wasm_bindgen(method, js_name = getRandomValues, catch)]
```

Comment:
You are forfeiting the check if `Crypto` is available though.
But if your intention was just to get rid of the nesting, how about prefacing the function like this:
```rust
if CRYPTO.with(Option::is_none) {
    return Err(Error::WEB_CRYPTO)
}

// use `get_random_values()` without nesting:
get_random_values(&buf);
```
Let me know how you want to proceed.

My recommendation is to keep the check in place, there's all sort of strange JS environments out there, `getrandom` is very foundational and should support these imo. And `CRYPTO` is cached after all so there is no performance overhead.

Reply:
Checking if the global `crypto` object exists doesn't really change what platforms/environments we support. It just changes what error message we get when the call to `crypto.getRandomValues` fails. Before, we were checking for existence of the object as part of our platform detection (and to cache the value in `RngSource`). Now that we don't have two different implementations for Web vs Node, we don't need to do any of that.

So if I'm understanding things correctly, having the check for the existence of the object just lets us return `Error::WEB_CRYPTO` to the user, informing them that `"Web Crypto API is unavailable"` instead of my implementation above which would make us return `Error::WEB_GET_RANDOM_VALUES` in all places.

I think returning good error messages is valuable, so I think keeping some sort of check is OK, provided that we aren't making an additional FFI call every time (which I don't think we do because the result of  `is_none` is cached).

I personally find the un-nested version easier to read, so we might try a version that avoids too much nesting (potentially by using a helper function).

Reply:
> Checking if the global `crypto` object exists doesn't really change what platforms/environments we support. It just changes what error message we get when the call to `crypto.getRandomValues` fails.

Ah, you are relying on the fact that WBG just catches any exception when tagging it with `catch`. While this is currently the case, its not intended to be used like this. Ideally, this would only catch exceptions thrown by the function. However, I guess this behavior has to stay in place because otherwise it would be a breaking change.

> I personally find the un-nested version easier to read, so we might try a version that avoids too much nesting (potentially by using a helper function).

Sure!

Reply:
Alright, let me know if you like this better!

Reply:
Maybe we can just add a separate check with caching?
```rust
const CRYPTO_PRESENT: u8 = 0;
const CRYPTO_NOT_PRESENT: u8 = 1;
const CRYPTO_UNINIT: u8 = 255;

static CRYPTO_CHECK: AtomicU8 = AtomicU8::new(CRYPTO_UNINIT);

pub fn fill_inner(dest: &mut [MaybeUninit<u8>]) -> Result<(), Error> {
    match CRYPTO_CHECK.load(Relaxed) {
        CRYPTO_PRESENT => {}
        CRYPTO_NOT_PRESENT => return Err(Error::WEB_CRYPTO),
        _ => {
            let global: Global = global().unchecked_into();
            let crypto = global.crypto();
            if crypto.is_object() {
                CRYPTO_CHECK.store(CRYPTO_PRESENT, Relaxed);
            } else {
                CRYPTO_CHECK.store(CRYPTO_NOT_PRESENT, Relaxed);
                return Err(Error::WEB_CRYPTO);
            }
        }
    }
    // ...
}
```
IIUC the atomic types are still present on "non-atomic" WASM targets.

Reply:
If the goal is to cache the check then that's not necessary. `CRYPTO` is already cached in a thread-local, its result resides in Rust memory after the first call and it makes no FFI calls afterwards.
Let me know if I misunderstood you.

Reply:
I meant it in the context of @josephlr's proposal to remove use of `thread_local_v2` and rely on `js_namespace` instead.

Reply:
Ah, I see!

From my understanding the issue was getting rid of the nesting, but that has been achieved, right? Is there some other motivation to get rid of `thread_local_v2`?

On that note: your proposal will store the check globally. But different threads may have different outcomes. This isn't hypothetical either, some worklets don't have access to `Crypto`. So this needs to be in a thread-local to work correctly.

Reply:
I am not familiar with the `wasm-bindgen` inner workings, but wouldn't `js_namespace`-based code result in a lighter-weight WASM file? Plus, the resulting source code will be slightly smaller (assuming that we do not care about the `WEB_CRYPTO` error).

Reply:
Okay, that fast: [according to the specification](https://www.w3.org/TR/2017/REC-WebCryptoAPI-20170126/#Crypto-method-getRandomValues), `Crypto.getRandomValues()` can only fail if we don't supply the right buffer type or we exceed the maximum byte length of 65536. So instead of getting rid of `WEB_CRYPTO` I will go ahead and get rid of `WEB_GET_RANDOM_VALUES`.

Let me know what you think!

Reply:
Done in https://github.com/rust-random/getrandom/pull/559/commits/8cf68328aa898966233496da384185744dd25d87.